<a href="https://colab.research.google.com/github/Rorando111/cpe019/blob/main/Celeste__Assignment_7_1_Classifications_and_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Technological Institute of the Philippines | Quezon City - Computer Engineering
--- | ---
Course Code: | CPE 019
Code Title: | Emerging Technologies in CpE 2
2nd Semester | AY 2023-2024
<hr> | <hr>
<u>**Assignment 7.1 : Classifications and Regression** |
**Name** | Celeste, Rolando
**Section** | CPE32S9
**Date Performed**: | 04/14/2024
**Date Submitted**: | 04/15/2024
**Instructor**: | Engr. Roman M. Richard

<hr>

Instructions:

1. Choose any dataset applicable to the classification problem, and also, choose any dataset applicable to the regression problem.
2. Explain your datasets and the problem being addressed.
3. For classification, do the following:
  1. Create a base model
  2. Evaluate the model with k-fold cross validation
  3. Improve the accuracy of your model by applying additional hidden layers
4. For regression, do the following:
  1. Create a base model
  2. Improve the model by standardizing the dataset
  3. Show tuning of layers and neurons (see evaluating small and larger networks)
5. Submit the link to your Google Colab (make sure that it is accessible to me)


### Datasets:

* apple_quality dataset

  - this dataset contains the characteristics of apples the are commonly distinguished to determine the apple's quality. These dataset aims to determine wether an apple is good or bad depending on the data that was gathered from them

* Student_Performance dataset
  - This dataset contains dail activities of students and the results from their previous examinations. This dataset aims to verify the factors that can affect a student's performance.

### **Classification Dataset**


### **Importing libraries and dataset to be used**

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix, precision_recall_curve, roc_auc_score, roc_curve, accuracy_score
from sklearn.ensemble import RandomForestClassifier

import seaborn as sns

%matplotlib inline

In [ ]:
from keras.models  import Sequential
from keras.layers import Input, Dense, Flatten, Dropout, BatchNormalization
from keras.optimizers import Adam, SGD, RMSprop

In [ ]:
data = pd.read_csv('/content/apple_quality.csv')
data.head()

A_id      Size    Weight  Sweetness  Crunchiness  Juiciness  Ripeness  \
0   0.0 -3.970049 -2.512336   5.346330    -1.012009   1.844900  0.329840   
1   1.0 -1.195217 -2.839257   3.664059     1.588232   0.853286  0.867530   
2   2.0 -0.292024 -1.351282  -1.738429    -0.342616   2.838636 -0.038033   
3   3.0 -0.657196 -2.271627   1.324874    -0.097875   3.637970 -3.413761   
4   4.0  1.364217 -1.296612  -0.384658    -0.553006   3.030874 -1.303849   

        Acidity Quality  
0  -0.491590483    good  
1  -0.722809367    good  
2   2.621636473     bad  
3   0.790723217    good  
4   0.501984036    good

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4001 entries, 0 to 4000
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   A_id         4000 non-null   float64
 1   Size         4000 non-null   float64
 2   Weight       4000 non-null   float64
 3   Sweetness    4000 non-null   float64
 4   Crunchiness  4000 non-null   float64
 5   Juiciness    4000 non-null   float64
 6   Ripeness     4000 non-null   float64
 7   Acidity      4001 non-null   object 
 8   Quality      4000 non-null   object 
dtypes: float64(7), object(2)
memory usage: 281.4+ KB


### Cleaning and processing data

- For this dataset, I had to remove unnecessary columns and change the target variable to binary so that it can fit into my model.

In [ ]:
data.dropna(inplace = True)

In [ ]:
data.drop(columns = ['A_id'], inplace = True)

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4000 entries, 0 to 3999
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Size         4000 non-null   float64
 1   Weight       4000 non-null   float64
 2   Sweetness    4000 non-null   float64
 3   Crunchiness  4000 non-null   float64
 4   Juiciness    4000 non-null   float64
 5   Ripeness     4000 non-null   float64
 6   Acidity      4000 non-null   object 
 7   Quality      4000 non-null   object 
dtypes: float64(6), object(2)
memory usage: 281.2+ KB


In [ ]:
data["Quality"] = data["Quality"].apply(lambda toLabel: 0 if toLabel == 'bad' else 1)

In [ ]:
data

Size    Weight  Sweetness  Crunchiness  Juiciness  Ripeness  \
0    -3.970049 -2.512336   5.346330    -1.012009   1.844900  0.329840   
1    -1.195217 -2.839257   3.664059     1.588232   0.853286  0.867530   
2    -0.292024 -1.351282  -1.738429    -0.342616   2.838636 -0.038033   
3    -0.657196 -2.271627   1.324874    -0.097875   3.637970 -3.413761   
4     1.364217 -1.296612  -0.384658    -0.553006   3.030874 -1.303849   
...        ...       ...        ...          ...        ...       ...   
3995  0.059386 -1.067408  -3.714549     0.473052   1.697986  2.244055   
3996 -0.293118  1.949253  -0.204020    -0.640196   0.024523 -1.087900   
3997 -2.634515 -2.138247  -2.440461     0.657223   2.199709  4.763859   
3998 -4.008004 -1.779337   2.366397    -0.200329   2.161435  0.214488   
3999  0.278540 -1.715505   0.121217    -1.154075   1.266677 -0.776571   

           Acidity  Quality  
0     -0.491590483        1  
1     -0.722809367        1  
2      2.621636473        0  
3      0.790723217        1  
4      0.501984036        1  
...            ...      ...  
3995   0.137784369        0  
3996   1.854235285        1  
3997  -1.334611391        0  
3998  -2.229719806        1  
3999   1.599796456        1  

[4000 rows x 8 columns]

### **Splitting train and test samples**

In [ ]:
X = data.iloc[:,:-1].values
Y = data.iloc[:,-1].values

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(X, Y, random_state = 1111, test_size = 0.25)

### **Scaling data**

In [ ]:
normalizer = StandardScaler()
X_train_norm = normalizer.fit_transform(x_train)
X_test_norm = normalizer.transform(x_test)

### **Creating base model**

In [ ]:
model_1 = Sequential()
model_1.add(Dense(12, input_shape=(7,), activation="relu"))
model_1.add(Dense(12, activation="relu"))
model_1.add(Dense(1, activation="sigmoid"))

model_1.compile(optimizer=SGD(learning_rate=0.001), loss='binary_crossentropy', metrics=['accuracy'])
run_hist_1 = model_1.fit(X_train_norm, y_train, validation_data=(X_test_norm, y_test), epochs=1000)

avg_accuracy = np.mean(run_hist_1.history['accuracy'])
print("Average Accuracy: {}%".format((avg_accuracy).round(3)*100))

avg_val_accuracy = np.mean(run_hist_1.history['val_accuracy'])
print("Average Validation Accuracy: {}%".format((avg_val_accuracy).round(3)*100))

Epoch 1/1000
94/94 [==============================] - 2s 9ms/step - loss: 0.6976 - accuracy: 0.6063 - val_loss: 0.6912 - val_accuracy: 0.6080
Epoch 2/1000
94/94 [==============================] - 0s 5ms/step - loss: 0.6909 - accuracy: 0.6287 - val_loss: 0.6847 - val_accuracy: 0.6150
Epoch 3/1000
94/94 [==============================] - 1s 14ms/step - loss: 0.6849 - accuracy: 0.6357 - val_loss: 0.6789 - val_accuracy: 0.6290
Epoch 4/1000
94/94 [==============================] - 1s 10ms/step - loss: 0.6796 - accuracy: 0.6433 - val_loss: 0.6738 - val_accuracy: 0.6360
Epoch 5/1000
94/94 [==============================] - 1s 9ms/step - loss: 0.6748 - accuracy: 0.6483 - val_loss: 0.6691 - val_accuracy: 0.6450
Epoch 6/1000
94/94 [==============================] - 1s 8ms/step - loss: 0.6705 - accuracy: 0.6523 - val_loss: 0.6649 - val_accuracy: 0.6590
Epoch 7/1000
94/94 [==============================] - 1s 7ms/step - loss: 0.6665 - accuracy: 0.6597 - val_loss: 0.6611 - val_accuracy: 0.6710
Epoc

Observation:
- The base model for this dataset resulted in 86.9% accuracy and 84.7% validation accuracy. Based on this results, we could say that the model is training well. However, this model can be further improved.

### **K-fold Cross Validation of base model**

In [ ]:
import tensorflow

In [ ]:
from sklearn.model_selection import KFold
from tensorflow.keras.callbacks import EarlyStopping

k = 5
kf = KFold(n_splits=k, random_state=None, shuffle=True)

fold_accuracies = []
train_loss = []
train_acc = []
val_loss = []
val_acc = []

for train_index, val_index in kf.split(X_train_norm):
    X_fold_train, X_fold_val = X_train_norm[train_index], X_train_norm[val_index]
    y_fold_train, y_fold_val = y_train[train_index], y_train[val_index]
    #Early stopping
    early_stopping = EarlyStopping(monitor='val_loss', patience=15, restore_best_weights=True)
    #Train model
    run_hist_fold = model_1.fit(X_fold_train, y_fold_train, validation_data=(X_fold_val, y_fold_val), epochs=1000, callbacks=[early_stopping])
    if early_stopping.stopped_epoch:
        print(f"Early stopping at epoch {early_stopping.stopped_epoch}")
    fold_accuracy = run_hist_fold.history['accuracy'][-1]
    train_loss.append(run_hist_fold.history['loss'][0])
    train_acc.append(run_hist_fold.history['accuracy'][0])
    val_loss.append(run_hist_fold.history['val_loss'][-1])
    val_acc.append(run_hist_fold.history['val_accuracy'][-1])
    fold_accuracies.append(fold_accuracy)

#Print the average accuracies of results
print("Mean fold accuracy: {}%".format(np.mean(fold_accuracies).round(3)*100))
print("Mean train loss: {}%".format(np.mean(train_loss).round(3)*100))
print("Mean train accuracy: {}%".format(np.mean(np.mean(train_acc)).round(3)*100))
print("Mean val loss: {}%".format(np.mean(np.mean(val_loss)).round(3)*100))
print("Mean val accuracy: {}%".format(np.mean(np.mean(val_acc)).round(3)*100))

Epoch 1/1000
75/75 [==============================] - 0s 3ms/step - loss: 0.2516 - accuracy: 0.8983 - val_loss: 0.2410 - val_accuracy: 0.9017
Epoch 2/1000
75/75 [==============================] - 0s 4ms/step - loss: 0.2515 - accuracy: 0.8983 - val_loss: 0.2408 - val_accuracy: 0.9017
Epoch 3/1000
75/75 [==============================] - 0s 3ms/step - loss: 0.2514 - accuracy: 0.8979 - val_loss: 0.2406 - val_accuracy: 0.9017
Epoch 4/1000
75/75 [==============================] - 0s 3ms/step - loss: 0.2513 - accuracy: 0.8988 - val_loss: 0.2405 - val_accuracy: 0.9000
Epoch 5/1000
75/75 [==============================] - 0s 3ms/step - loss: 0.2512 - accuracy: 0.8975 - val_loss: 0.2404 - val_accuracy: 0.8983
Epoch 6/1000
75/75 [==============================] - 0s 4ms/step - loss: 0.2511 - accuracy: 0.8975 - val_loss: 0.2403 - val_accuracy: 0.8983
Epoch 7/1000
75/75 [==============================] - 0s 3ms/step - loss: 0.2510 - accuracy: 0.8975 - val_loss: 0.2403 - val_accuracy: 0.8983
Epoch 

**Analysis:**
- I used 5 folds for the cross validation, with an average fold accuracy of 92.2%. An average of 92.2% for the performance evaluation of this model can be further improved. In addidtion, I decided to use the early stopping function because I have observed from my earlier testing that the validation was overfitting.

### **Improving Accuracy**

In [ ]:
import tensorflow

In [ ]:
from tensorflow.keras.callbacks import EarlyStopping

# Define the model
model_2 = Sequential()
model_2.add(Dense(20, input_shape=(7,), activation='relu'))
model_2.add(Dense(20, activation='relu'))
model_2.add(Dense(20, activation='relu'))
model_2.add(Dense(1, activation='sigmoid'))

early_stopping = EarlyStopping(monitor='val_loss', patience=15, restore_best_weights=True)

model_2.compile(optimizer=SGD(learning_rate=0.03), loss='binary_crossentropy', metrics=['accuracy'])

run_hist_2 = model_2.fit(X_train_norm, y_train, validation_data=(X_test_norm, y_test), epochs=1500, callbacks=[early_stopping])

avg_accuracy = np.mean(run_hist_2.history['accuracy'])
print("Average Accuracy: {}%".format((avg_accuracy).round(3)*100))

avg_val_accuracy = np.mean(run_hist_2.history['val_accuracy'])
print("Average Validation Accuracy: {}%".format((avg_val_accuracy).round(3)*100))

Epoch 1/1500
94/94 [==============================] - 3s 7ms/step - loss: 0.6451 - accuracy: 0.6490 - val_loss: 0.6044 - val_accuracy: 0.7120
Epoch 2/1500
94/94 [==============================] - 0s 5ms/step - loss: 0.5596 - accuracy: 0.7430 - val_loss: 0.5214 - val_accuracy: 0.7720
Epoch 3/1500
94/94 [==============================] - 0s 3ms/step - loss: 0.4821 - accuracy: 0.7953 - val_loss: 0.4538 - val_accuracy: 0.8060
Epoch 4/1500
94/94 [==============================] - 0s 3ms/step - loss: 0.4260 - accuracy: 0.8137 - val_loss: 0.4093 - val_accuracy: 0.8150
Epoch 5/1500
94/94 [==============================] - 0s 3ms/step - loss: 0.3874 - accuracy: 0.8297 - val_loss: 0.3786 - val_accuracy: 0.8320
Epoch 6/1500
94/94 [==============================] - 0s 3ms/step - loss: 0.3585 - accuracy: 0.8430 - val_loss: 0.3566 - val_accuracy: 0.8380
Epoch 7/1500
94/94 [==============================] - 0s 3ms/step - loss: 0.3371 - accuracy: 0.8500 - val_loss: 0.3386 - val_accuracy: 0.8460
Epoch 

**Analysis:**
- For the modification of the base model, I decided to add another hidden layer with 20 nodes and an activation of 'relu'. I also decided to use early stopping for this model in order to  prevent overfitting, but it seems like there wasn't any need. Observing the accuracy results, with an average accuracy of 92% and average validation accuracy of 89.9%, this model clearly shows better performance compared to the base model.

In [ ]:
from sklearn.model_selection import KFold
from tensorflow.keras.callbacks import EarlyStopping

k = 5
kf = KFold(n_splits=k, random_state=None, shuffle=True)


fold_accuracies = []
train_loss = []
train_acc = []
val_loss = []
val_acc = []

for train_index, val_index in kf.split(X_train_norm):
    X_fold_train, X_fold_val = X_train_norm[train_index], X_train_norm[val_index]
    y_fold_train, y_fold_val = y_train[train_index], y_train[val_index]

    early_stopping = EarlyStopping(monitor='val_loss', patience=15, restore_best_weights=True)

    run_hist_fold = model_2.fit(X_fold_train, y_fold_train, validation_data=(X_fold_val, y_fold_val), epochs=1500, callbacks=[early_stopping])
    if early_stopping.stopped_epoch:
        print(f"Early stopping at epoch {early_stopping.stopped_epoch}")

    fold_accuracy = run_hist_fold.history['accuracy'][-1]
    train_loss.append(run_hist_fold.history['loss'][0])
    train_acc.append(run_hist_fold.history['accuracy'][0])
    val_loss.append(run_hist_fold.history['val_loss'][-1])
    val_acc.append(run_hist_fold.history['val_accuracy'][-1])
    fold_accuracies.append(fold_accuracy)



print("Mean fold accuracy: {}%".format(np.mean(fold_accuracies).round(3)*100))
print("Mean train loss: {}%".format(np.mean(train_loss).round(3)*100))
print("Mean train accuracy: {}%".format(np.mean(np.mean(train_acc)).round(3)*100))
print("Mean val loss: {}%".format(np.mean(np.mean(val_loss)).round(3)*100))
print("Mean val accuracy: {}%".format(np.mean(np.mean(val_acc)).round(3)*100))

Epoch 1/1500
75/75 [==============================] - 0s 4ms/step - loss: 0.1137 - accuracy: 0.9583 - val_loss: 0.1109 - val_accuracy: 0.9583
Epoch 2/1500
75/75 [==============================] - 0s 4ms/step - loss: 0.1126 - accuracy: 0.9563 - val_loss: 0.1146 - val_accuracy: 0.9550
Epoch 3/1500
75/75 [==============================] - 0s 3ms/step - loss: 0.1098 - accuracy: 0.9617 - val_loss: 0.1315 - val_accuracy: 0.9533
Epoch 4/1500
75/75 [==============================] - 0s 4ms/step - loss: 0.1098 - accuracy: 0.9608 - val_loss: 0.1213 - val_accuracy: 0.9500
Epoch 5/1500
75/75 [==============================] - 0s 4ms/step - loss: 0.1057 - accuracy: 0.9633 - val_loss: 0.1343 - val_accuracy: 0.9500
Epoch 6/1500
75/75 [==============================] - 0s 4ms/step - loss: 0.1103 - accuracy: 0.9583 - val_loss: 0.1371 - val_accuracy: 0.9433
Epoch 7/1500
75/75 [==============================] - 0s 4ms/step - loss: 0.1060 - accuracy: 0.9608 - val_loss: 0.1357 - val_accuracy: 0.9500
Epoch 

**Analysis:**
- The k-fold cross validation also concludes that this model performs better and has a high accuracy.

## **Regression**

### **Importing libraries and dataset to be used**

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import confusion_matrix, precision_recall_curve, roc_auc_score, roc_curve, accuracy_score
from sklearn.ensemble import RandomForestClassifier

import seaborn as sns

%matplotlib inline

In [ ]:
from keras.models  import Sequential
from keras.layers import Input, Dense, Flatten, Dropout, BatchNormalization
from keras.optimizers import Adam, SGD, RMSprop

In [ ]:
df = pd.read_csv('/content/Student_Performance.csv')

### **Cleaning and processing data**

In [ ]:
df.head()

Hours Studied  Previous Scores Extracurricular Activities  Sleep Hours  \
0              7               99                        Yes            9   
1              4               82                         No            4   
2              8               51                        Yes            7   
3              5               52                        Yes            5   
4              7               75                         No            8   

   Sample Question Papers Practiced  Performance Index  
0                                 1               91.0  
1                                 2               65.0  
2                                 2               45.0  
3                                 2               36.0  
4                                 5               66.0

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 6 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Hours Studied                     10000 non-null  int64  
 1   Previous Scores                   10000 non-null  int64  
 2   Extracurricular Activities        10000 non-null  object 
 3   Sleep Hours                       10000 non-null  int64  
 4   Sample Question Papers Practiced  10000 non-null  int64  
 5   Performance Index                 10000 non-null  float64
dtypes: float64(1), int64(4), object(1)
memory usage: 468.9+ KB


In [ ]:
df["Extracurricular Activities"] = df["Extracurricular Activities"].apply(lambda toLabel: 0 if toLabel == 'No' else 1)

In [ ]:
df.head()

Hours Studied  Previous Scores  Extracurricular Activities  Sleep Hours  \
0              7               99                           1            9   
1              4               82                           0            4   
2              8               51                           1            7   
3              5               52                           1            5   
4              7               75                           0            8   

   Sample Question Papers Practiced  Performance Index  
0                                 1               91.0  
1                                 2               65.0  
2                                 2               45.0  
3                                 2               36.0  
4                                 5               66.0

### **Splitting train and test samples**

In [ ]:
x = df.iloc[:,:-1].values
y = df.iloc[:,-1].values

In [ ]:
X_train, X_test, Y_train, Y_test = train_test_split(x, y, random_state = 1111, test_size = 0.25)

### **Creating base model**

In [ ]:
model = Sequential()
model.add(Dense(20, input_shape=(5,), activation='relu'))
model.add(Dense(1))

model.compile(optimizer="rmsprop", loss="mse", metrics=["mae"])
hist_3 = model.fit(X_train, Y_train, epochs=100, batch_size=32, validation_data=(X_test, Y_test))

# Print the MSE for the training and testing sets
print("Base model MSE:")
print("Average Training MSE: ", hist_3.history['loss'][-1])
print("Average Validation MAE: ", hist_3.history['val_loss'][-1])

Epoch 1/100
235/235 [==============================] - 2s 5ms/step - loss: 345.3213 - mae: 12.9275 - val_loss: 71.9592 - val_mae: 6.8981
Epoch 2/100
235/235 [==============================] - 1s 4ms/step - loss: 52.1946 - mae: 5.8676 - val_loss: 41.6828 - val_mae: 5.1656
Epoch 3/100
235/235 [==============================] - 1s 3ms/step - loss: 35.8634 - mae: 4.8249 - val_loss: 31.9158 - val_mae: 4.5271
Epoch 4/100
235/235 [==============================] - 1s 3ms/step - loss: 30.4686 - mae: 4.4392 - val_loss: 34.3174 - val_mae: 4.7404
Epoch 5/100
235/235 [==============================] - 1s 2ms/step - loss: 28.3563 - mae: 4.2950 - val_loss: 26.9682 - val_mae: 4.1646
Epoch 6/100
235/235 [==============================] - 1s 3ms/step - loss: 26.6918 - mae: 4.1710 - val_loss: 25.9089 - val_mae: 4.0800
Epoch 7/100
235/235 [==============================] - 1s 3ms/step - loss: 24.9774 - mae: 4.0233 - val_loss: 24.6353 - val_mae: 3.9745
Epoch 8/100
235/235 [==============================] 

**Analysis:**
- For the base model's architecture, it only consists of 1 hidden layer and 20 nodes. The training process was for 100 epochs, with a batch size of 32, using 'rmsprop' as optimizer and mse as loss function. While the testing MSE fluctuated but generally showed a declining trend, the training MSE showed a consistent decline throughout the course of the epochs. The Average training MSE was 4.628, and the Average Valdidation MAE of 4.825. This implies that although the model fits the data rather well, it could be a little overfitting. It could take further assessment to improve the model's functionality.

### **Scaling data**

In [ ]:
normalizer = StandardScaler()
X_train_scaled = normalizer.fit_transform(X_train)
X_test_scaled = normalizer.transform(X_test)

In [ ]:
import tensorflow as tf

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

model_1 = Sequential()
model_1.add(Dense(12, input_shape=(5,), activation="relu"))
model_1.add(Dense(12, activation="relu"))
model_1.add(Dense(1, activation="linear"))

model_1.compile(optimizer="rmsprop", loss="mse", metrics=["mae"])
# Train the model
hist_4 = model_1.fit(X_train_scaled, Y_train, epochs=100, batch_size=32, validation_data=(X_test_scaled, Y_test))
# Print the MSE for the training and testing sets
print("Scaled model MSE:")
print("Average Training MSE: ", hist_4.history['loss'][-1])
print("Average Validation MAE: ", hist_4.history['val_loss'][-1])

Epoch 1/100
235/235 [==============================] - 2s 6ms/step - loss: 3180.7478 - mae: 52.9499 - val_loss: 2690.1562 - val_mae: 48.2481
Epoch 2/100
235/235 [==============================] - 1s 5ms/step - loss: 2067.3293 - mae: 41.3675 - val_loss: 1204.4106 - val_mae: 30.6045
Epoch 3/100
235/235 [==============================] - 1s 5ms/step - loss: 549.3867 - mae: 18.8392 - val_loss: 117.6543 - val_mae: 8.8222
Epoch 4/100
235/235 [==============================] - 1s 4ms/step - loss: 55.4483 - mae: 5.9058 - val_loss: 38.6720 - val_mae: 4.9953
Epoch 5/100
235/235 [==============================] - 1s 4ms/step - loss: 30.4201 - mae: 4.3829 - val_loss: 26.5734 - val_mae: 4.1299
Epoch 6/100
235/235 [==============================] - 1s 4ms/step - loss: 21.5124 - mae: 3.6830 - val_loss: 19.1234 - val_mae: 3.4882
Epoch 7/100
235/235 [==============================] - 1s 5ms/step - loss: 16.1829 - mae: 3.1985 - val_loss: 14.7122 - val_mae: 3.0047
Epoch 8/100
235/235 [===================

### **Tuning using Larger model**

In [ ]:
# Try a larger model
model_2 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(24, activation='relu', input_shape=(X_train_scaled.shape[1],)),
    tf.keras.layers.Dense(24, activation='relu'),
    tf.keras.layers.Dense(24, activation='relu'),
    tf.keras.layers.Dense(16, activation='linear'),
    tf.keras.layers.Dense(1)
])

# Compile the model
model_2.compile(optimizer="adam", loss="mse", metrics=["mse"])

# Train the model
hist_L = model_2.fit(X_train_scaled, Y_train, epochs=100, batch_size=32, validation_data=(X_test_scaled, Y_test))

# Print the MSE for the training and testing sets
print("Larger model MSE:")
print("Training: ", hist_L.history['loss'][-1])
print("Testing: ", hist_L.history['val_loss'][-1])

Epoch 1/100
235/235 [==============================] - 6s 16ms/step - loss: 1038.1594 - mse: 1038.1594 - val_loss: 51.6828 - val_mse: 51.6828
Epoch 2/100
235/235 [==============================] - 3s 11ms/step - loss: 32.5972 - mse: 32.5972 - val_loss: 21.2618 - val_mse: 21.2618
Epoch 3/100
235/235 [==============================] - 1s 4ms/step - loss: 13.7497 - mse: 13.7497 - val_loss: 9.5131 - val_mse: 9.5131
Epoch 4/100
235/235 [==============================] - 1s 3ms/step - loss: 7.6584 - mse: 7.6584 - val_loss: 7.1206 - val_mse: 7.1206
Epoch 5/100
235/235 [==============================] - 1s 3ms/step - loss: 6.2954 - mse: 6.2954 - val_loss: 6.7191 - val_mse: 6.7191
Epoch 6/100
235/235 [==============================] - 1s 3ms/step - loss: 5.7174 - mse: 5.7174 - val_loss: 5.9126 - val_mse: 5.9126
Epoch 7/100
235/235 [==============================] - 1s 3ms/step - loss: 5.4412 - mse: 5.4412 - val_loss: 5.6489 - val_mse: 5.6489
Epoch 8/100
235/235 [==============================] 

### **Tuning using Smaller model**

In [ ]:
# Try a smaller model
model_3 = tf.keras.models.Sequential([
    tf.keras.layers.Dense(8, activation='relu', input_shape=(X_train_scaled.shape[1],)),
    tf.keras.layers.Dense(4, activation='relu'),
    tf.keras.layers.Dense(1, activation='linear')
])

# Compile the model
model_3.compile(optimizer='adam', loss='mse')

# Train the model
hist_S = model_3.fit(X_train_scaled, Y_train, epochs=100, batch_size=32, validation_data=(X_test_scaled, Y_test))

print("Smaller model MSE:")
print("Training: ", hist_S.history['loss'][-1])
print("Testing: ", hist_S.history['val_loss'][-1])

Epoch 1/100
235/235 [==============================] - 1s 3ms/step - loss: 3318.6548 - val_loss: 3046.5125
Epoch 2/100
235/235 [==============================] - 0s 2ms/step - loss: 2824.7280 - val_loss: 2347.1494
Epoch 3/100
235/235 [==============================] - 1s 2ms/step - loss: 1875.7122 - val_loss: 1248.5487
Epoch 4/100
235/235 [==============================] - 1s 3ms/step - loss: 791.1199 - val_loss: 378.7841
Epoch 5/100
235/235 [==============================] - 1s 3ms/step - loss: 198.1167 - val_loss: 94.0139
Epoch 6/100
235/235 [==============================] - 1s 3ms/step - loss: 66.5021 - val_loss: 54.4953
Epoch 7/100
235/235 [==============================] - 1s 3ms/step - loss: 47.7359 - val_loss: 44.4430
Epoch 8/100
235/235 [==============================] - 1s 3ms/step - loss: 40.2208 - val_loss: 37.9055
Epoch 9/100
235/235 [==============================] - 1s 2ms/step - loss: 34.6242 - val_loss: 32.5974
Epoch 10/100
235/235 [==============================] - 1s

### **Conclusion;**

- In this activity, i learned how different models are used on different types of datasets. I also learned and was also convinced that the accuracy of training the model on a dataset depends on the parameters you have set for the model.